File: haskell/glue/src/Glue/Lib/List/Member.hs
module Glue.Lib.List.Member where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

member :: IR Eval
member = NativeFunc memberImpl

memberImpl :: IR Eval -> Eval (IR Eval)
memberImpl itemIR = pure $ NativeFunc (memberIn itemIR)

memberIn :: IR Eval -> IR Eval -> Eval (IR Eval)
memberIn item list = case list of
    List xs -> pure . Bool $ item `elem` xs
    _ -> throwError $ wrongArgumentType ["list"]

File: dart/glue/lib/src/lib/list/member.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Member function - checks if an item is in a list
/// Mirrors Haskell Glue.Lib.List.Member.member exactly
Ir member = IrNativeFunc(memberImpl);

/// Member function implementation
/// Mirrors Haskell Glue.Lib.List.Member.memberImpl exactly
Eval<Ir> memberImpl(Ir itemIr) {
  return Eval.pure(IrNativeFunc(memberIn(itemIr)));
}

/// Helper function for list argument
/// Mirrors Haskell Glue.Lib.List.Member.memberIn exactly
Eval<Ir> Function(Ir) memberIn(Ir item) {
  return (Ir list) => switch (list) {
    IrList(:final elements) => Eval.pure(IrBool(elements.contains(item))),
    _ => throwError(wrongArgumentType(['list'])),
  };
}
